# Introduction

*Sprocket Central Pty Ltd , a medium size bikes & cycling accessories organisation. They're looking to expand into a new market, they want to know what is the suitable marketing strategy.Our task is to analyse the datasets to help Sprocket Central Pty Ltd grow its business.The organisation has a large dataset relating to its customers* 

*The company provided  3 datasets:*
* Customer Demographic 
* Customer Addresses
* Transactions data in the past 3 months 

*Our task is to review the quality of the dataset to ensure that it is ready for our analysis part and take note of any assumptions or issues we need as well as recommendations going forward to mitigate current data quality concerns.*

#### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns 
import matplotlib.pyplot as plt
import plotly.express as px

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_columns', None)

#### Loading the dataset

In [3]:
transaction = pd.read_excel("D:\Downloads\data\Transaction.xlsx")
customer_address = pd.read_excel("D:\Downloads\data\CustomerAddress.xlsx")
customer_demographic = pd.read_excel("D:\Downloads\data\CustomerDemographic.xlsx")

# Data Quality Assessment

##### We will asses the dataset one by one

##### 1. Transactions
##### 2. Customer Addresses
##### 3. Customer Demographic

## Transaction Data

In [5]:
# Display the first few rows

display(transaction.head())
print("Size: ",transaction.shape)

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

Size:  (20000, 13)


In [6]:
# Function to identify missing values

def check_null_values(df):
    null_count = df.isnull().sum().sort_values(ascending=False)
    null_proportion = ((df.isnull().sum() / df.isnull().count()*100).sort_values(ascending = False))
    null_df = pd.concat([null_count, null_proportion], keys = ['Total_no_of_Null', 'Total_%_of_Null'],axis=1)
    return null_df

In [7]:
# Check for missing values 
check_null_values(transaction)

Total_no_of_Null  Total_%_of_Null
online_order                          360            1.800
brand                                 197            0.985
product_line                          197            0.985
product_class                         197            0.985
product_size                          197            0.985
standard_cost                         197            0.985
product_first_sold_date               197            0.985
transaction_id                          0            0.000
product_id                              0            0.000
customer_id                             0            0.000
transaction_date                        0            0.000
order_status                            0            0.000
list_price                              0            0.000

In [8]:
# Check Duplicates

if transaction.duplicated().any():
    print("Duplicates found!")
else:
    print("No duplicates found.")

No duplicates found.


In [9]:
# Check summary information about the DataFrame
transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [10]:
# Statistical summary of the data
transaction.describe().T.style.background_gradient(cmap='turbo')

In [11]:
# Distinct Customer IDs
transaction['customer_id'].nunique()

3494

### Inference:

* No. of records 20000
* Distinct Customer IDs 3494
* The Transaction dataset has missing values. the highest proportion of missing values is 1.8% in 'online_order' feature.
* 'brand', 'product_line', 'product_class', 'product_size', 'standard_cost', 'product_first_sold_date' contain missing values

##  Customer Addresses Data

In [12]:
# Display the first few rows

display(customer_address.head())
print("Size: ",customer_address.shape)

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

Size:  (3999, 6)


In [13]:
# Check for missing values 
check_null_values(customer_address)

Total_no_of_Null  Total_%_of_Null
customer_id                        0              0.0
address                            0              0.0
postcode                           0              0.0
state                              0              0.0
country                            0              0.0
property_valuation                 0              0.0

In [14]:
# Check Duplicates

if customer_address.duplicated().any():
    print("Duplicates found!")
else:
    print("No duplicates found.")

No duplicates found.


In [15]:
# Check summary information about the DataFrame
customer_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [16]:
# Statistical summary of the data
customer_address.describe().T.style.background_gradient(cmap='turbo')

In [17]:
# Distinct Customer IDs
customer_address['customer_id'].nunique()

3999

### Inference:

* No. of records 3999
* Distinct Customer IDs 3499
* No missing value found

## Customer Demographic Data

In [18]:
# Display the first few rows

display(customer_demographic.head())
print("Size: ",customer_demographic.shape)

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

Size:  (4000, 13)


In [19]:
# Check for missing values 
check_null_values(customer_demographic)

Total_no_of_Null  Total_%_of_Null
job_industry_category                             656           16.400
job_title                                         506           12.650
default                                           302            7.550
last_name                                         125            3.125
DOB                                                87            2.175
tenure                                             87            2.175
customer_id                                         0            0.000
first_name                                          0            0.000
gender                                              0            0.000
past_3_years_bike_related_purchases                 0            0.000
wealth_segment                                      0            0.000
deceased_indicator                                  0            0.000
owns_car                                            0            0.000

In [20]:
# Check Duplicates

if customer_demographic.duplicated().any():
    print("Duplicates found!")
else:
    print("No duplicates found.")

No duplicates found.


In [21]:
# Check summary information about the DataFrame
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [22]:
# Statistical summary of the data
customer_demographic.describe().T.style.background_gradient(cmap='turbo')

In [23]:
# Distinct Customer IDs
customer_demographic['customer_id'].nunique()

4000

### Inference:

* No. of records 4000
* Unique Customer IDs 4000
* Missing values found. highest proportion of missing value is 16.4% in 'job_industry_category' feature
* 'job_title', 'default', 'last_name', 'DOB', 'tenure' contain missing values

## Identified the data quality issues

* Missing values found in Transactions & Customer Demographic data but not in Customer Addresses data
* Missing values found in datasets, various feature such as 'online_order','brand', 'product_line', 'product_class', 'product_size', 'standard_cost', 'product_first_sold_date', 'job_industry_category','job_title', 'default', 'last_name', 'DOB', 'tenure' contain missing values


mitigation :  Start by thoroughly examining the datasets to identify the presence and extent of missing values. Document the specific variables or attributes where missing values are observed. Evaluate the data collection process to identify any potential issues that may have contributed to missing values.


_